src/jwtauthmiddleware-test-server.ts

```diff
@@ -0,0 +1,133 @@
+import express, { Request, Response } from 'express'
+import http from 'http'
+
+import { PublicKey } from './index'
+import { JwtAuthMiddleware } from './jwtauthmiddleware'
+import { JwtVerifyError } from './jwtverifyerror'
+import { ecPublicKey } from './testresources'
+
+const pubKeys: Record<string, Record<string, string | PublicKey>> = {
+  'http://localhost/oauth/token': {
+    '1@ES256': ecPublicKey
+  },
+  'http://localhost/oauth/token/moreexpires': {
+    '1@ES256': {
+      publicKey: ecPublicKey,
+      expiresSkew: 600
+    }
+  }
+}
+
+const revokedTokens = {
+  jtiRevoked: {
+    jti: 'jtiRevoked',
+    revokedAt: new Date('2023-02-03')
+  }
+}
+
+const audiences = ['http://localhost/']
+
+interface ServerConfig {
+  port: number
+  requestTimeout?: number
+}
+
+export class JwtAuthMiddlewareServer {
+  private port: number
+  private app: express.Express
+  private server?: http.Server
+
+  public constructor(options: ServerConfig) {
+    this.port = options.port
+    this.app = express()
+    this.app.use(express.json())
+
+    this.app.use(
+      '/mapped',
+      JwtAuthMiddleware(
+        pubKeys,
+        revokedTokens,
+        audiences,
+        (user: Record<string, string> & { body: Record<string, string> }) => {
+          if (user.issuer === 'http://localhost/oauth/token') {
+            // Map claims
+            user.eMail = user.body.email
+          }
+        }
+      )
+    )
+
+    this.app.use(
+      '/async',
+      JwtAuthMiddleware(
+        pubKeys,
+        revokedTokens,
+        audiences,
+        (user: Record<string, string> & { body: Record<string, string> }) => {
+          if (user.subject === 'error') {
+            return Promise.reject(new JwtVerifyError('Async error'))
+          } else {
+            return Promise.resolve('test')
+          }
+        }
+      )
+    )
+
+    this.app.use(
+      '/anonymous',
+      JwtAuthMiddleware(pubKeys, revokedTokens, audiences, null, {
+        allowAnonymous: true
+      })
+    )
+    this.app.use('/', JwtAuthMiddleware(pubKeys, revokedTokens, audiences))
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    this.app.use((err: Error, req: Request, res: Response, _next: (err?: Error | null) => void) => {
```

Comment:
```suggestion
    this.app.use((err: Error, req: Request, res: Response) => {
```

Reply:
When defining an error handler you need all four arguments. I think this is just how express knows that it's an error handler middleware function that's defined. https://expressjs.com/en/guide/error-handling.html

Reply:
I tried removing it on your branch and I'm not getting any errors, though. The linked documentation also shows examples where `next` is not provided. Are you getting errors?

Reply:
If you run `jwtauthmiddleware.test.ts` after removing `next` it throws the wrong error for the tests. It throws a 500 internal server error from axios which is caught in the `DefaultHttpRequestHandler`. When `next` is left in a 401 error is returned with the correct reason context for why the error was thrown.

Reply:
I'm still trying to figure out why exactly this is happening. I could have set up the server to respond weird but that's the behavior I'm seeing currently.